:sparkles: crd conversion webhook implementation

droot opened the pull request:
This PR implements CRD Conversion Webhook handler.

The design has been documented [here](https://docs.google.com/document/d/1AbXhGhQduyM1xextudw_hXZKfAACRqEOvBxh61pFMaU/edit?usp=sharing).

Note:
Need to add more tests. I am in process of converting the design doc in markdown format (doc link shared above)

droot commented:
/cc @shawn-hurley @JoelSpeed

shawn-hurley commented:
@droot did I miss how someone could move to a new hub version? It was not immediately clear to me how this would work. 

I think the overall simplification of the concept is really nice!  I need to read through the code still but the design doc is really good!



<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

DirectXMan12 commented:
@shawn-hurley I've never seen something that wasn't a little bit painful, but the process here would basically be:

1. Mark the new type as hub with the interface
2. Unmark the old type
3. Migrate your `ConvertTo` and `ConvertFrom` methods to the new hub
4. Add new conversions for your old hub

While this is a bit of a pain, it's not *more* painful (in my experience) that the k/k method, which, from my experience with HPA v2, looks a lot like:

1. Update the internal version to have a convenient representation for supporting the new external version
2. Generate the new generated conversions
3. Realize that the internal version looks a lot like the new external version, so now you need to rewrite your existing conversions to work with the new internal version
4. Write conversions for the old external version that used to look a lot like the internal version
5. (bonus round) realize that the new internal version needs to be slightly different from the external version for some reason (maybe it makes round-tripping a lot easier), so now you have to write conversions for your new external version too.

Now, that's not a high bar, *but* I also don't see a good way to make things better without supporting arbitrary conversion graphs and gradually growing a strange web of conversions.

coderanger commented on pkg/conversion/conversion.go:
```diff
@@ -0,0 +1,18 @@
+package conversion
```
Missing license header.

droot replied:
Added :)

coderanger commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,249 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
It isn't 2018 :-)

droot replied:
Fixed :)

droot commented on pkg/webhook/conversion/testData/pkg/apis/jobs/v1/externaljob_types.go:
```diff
@@ -0,0 +1,97 @@
+
+// since v1.ExternalJob is a spoke, it needs to be convertable.
+// It needs to implement convert to/from storage version
+// TODO(droot): evaluate advantages of taking `conversion.Hub` as input to
```
>Is there a good way to remove one of the two?
Annotations (we have started calling them markers now because annotation/tags are so overloaded in k8s world :)) is primarily needed for controller-gen to generate the manifests for conversion webhook. Only way to remove them would be to make controller-gen parse these types and examine if they implement these methods (which can be done, but is a bit stretch looking at current state of the code in controller-gen). Till then, this is the only option.

>I had kinda expected ConvertTo and ConvertFrom to take a runtime.Object initially but I concluded that it's probably a disadvantage as you lose some compile time checks theoretically
Agreed. using `Hub` has type-safety as well readability from an end-user's perspective. So will took out that TODO.

droot commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,249 @@
+	if err != nil {
+		return err
+	}
+	// shall we get Hub for dst type as well and ensure hubs are same ?
```
I added a check at the beginning of the function to ensure both src and dst belong to same API group to ensure src/dst have same hub type.

droot commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,249 @@
+		}
+	}
+
+	// neigher src nor dst are Hub, means both of them are spoke, so lets get the hub
```
Fixed.

droot commented:
@JoelSpeed 

> I'm so glad you've started work on this @droot as I need to bump a kubebuilder API version soon and need this 😅 Let me know if there's any way I can help with this

Awesome. Thanks for reviewing the PR. Usability feedback will be super important and I am glad that you will be using it your project, so keep us posted from end-user's perspective. Have addressed the comments and added a few more tests.

droot commented:
@shawn-hurley 

> @droot did I miss how someone could move to a new hub version? It was not immediately clear to me how this would work.

Pretty much along the lines as @DirectXMan12 described above. I will add those to the design doc.

mengqiy commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,250 @@
+	var hub conversion.Hub
+	var isHub, hubFoundAlready bool
+	for _, gvk := range gvks {
+		o, _ := wh.scheme.New(gvk)
```
The error should be checked here. If the gvk is not registered in the scheme, `scheme.New` will return an error.

droot replied:
Good pt. Added catch.

mengqiy commented on pkg/webhook/conversion/decoder.go:
```diff
@@ -0,0 +1,30 @@
+}
+
+// Decode decodes the inlined object.
+func (d *Decoder) Decode(content []byte) (runtime.Object, *schema.GroupVersionKind, error) {
```
I feel we can move [the decoder in admission pkg](https://github.com/kubernetes-sigs/controller-runtime/blob/8d94f663b1f552f74805cd8c44a1e03387f5a5d2/pkg/webhook/admission/decode.go#L44) to a shared the place and reuse it potentially.

droot replied:
Added a TODO for now. I am sure `conversion` pkg will evolve as it goes to `Beta` and we can reevaluate it at that point.

DirectXMan12 replied:
not convinced we need this at all (see above comment)

mengqiy commented on pkg/webhook/conversion/response.go:
```diff
@@ -0,0 +1,34 @@
+func errored(err error) *apix.ConversionResponse {
+	return &apix.ConversionResponse{
+		Result: metav1.Status{
+			Code:    http.StatusOK,
```
It seems to be wrong here.

droot replied:
Good catch, tool it out.

mengqiy commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,250 @@
+		return
+	}
+
+	// TODO(droot): may be move the conversion logic to a separate module to
```
IMO it's a good idea and align with what admission package. It may look like
```
func (wh *Webhook) ServeConvert(req ConversionRequest, resp ConversionResponse) {
  // Do something
}

func (wh *Webhook) ServeHTTP(w http.ResponseWriter, r *http.Request) {
  // Decoding http request to get ConversionRequest

  // Do conversion
  wh.ServeConvert(convReq, convResp)

  // Write the conversion response to http.ResponseWriter
}
```

droot replied:
I want to keep this as TODO for now and revisit later. The good thing it can be done as internal refactoring later.

droot commented:
/cc @mbohlool

droot commented:
@mbohlool  @mengqiy @JoelSpeed This is good to go from my perspective. I think this is a good starting point to get feedback for the CRD conversion feature.

droot commented:
>Wonder if it would make sense to add tests testing the src and dest are not convertible or if src is the hub and if dest is the hub and make sure each happy path is covered?

I just added a test to cover the case where hub is not defined. One test which is missing now is (spoke-to-spoke conversion), I will have to define a two API group under `testData` to do that.

droot commented on pkg/webhook/conversion/conversion_suite_test.go:
```diff
@@ -0,0 +1,38 @@
+
+func TestConversionWebhook(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecsWithDefaultAndCustomReporters(t, "application Suite", []Reporter{envtest.NewlineReporter{}})
```
Done.

droot commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,256 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	"encoding/json"
```
Done.

droot commented on pkg/webhook/conversion/conversion_test.go:
```diff
@@ -0,0 +1,192 @@
+
+	It("should convert objects successfully", func() {
+
+		v1Obj := &jobsv1.ExternalJob{
```
I know it looks more verbose in current form, but it makes the test more readable from perspective of `input` and `output` and expectations.

DirectXMan12 replied:
I'd just move it to outside, or at least make a helper -- it'll make it *way* easier to refactor later.

DirectXMan12 commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,258 @@
+	convertReview := &apix.ConversionReview{}
+	// TODO(droot): figure out if we want to split decoder for conversion
+	// request from the objects contained in the request
+	err = wh.decoder.DecodeInto(body, convertReview)
```
I'm not convinced we *actually* need this at all -- we know the object that we're decoding into, so there's nothing special to be done.  The only thing this buys us is potential future-proofing for proto, but there are other spots in this chain that use JSON directly, so ¯\\_(ツ)_/¯

DirectXMan12 commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,258 @@
+	srcGVK := src.GetObjectKind().GroupVersionKind()
+	dstGVK := dst.GetObjectKind().GroupVersionKind()
+
+	if srcGVK.GroupKind().String() != dstGVK.GroupKind().String() {
```
you should be able to compare these directly w/o converting to string first, IIRC

DirectXMan12 commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,258 @@
+
+	srcIsHub, dstIsHub := isHub(src), isHub(dst)
+	srcIsConvertible, dstIsConvertible := isConvertible(src), isConvertible(dst)
+
```
```suggestion
switch {
case srcIsHub && dstIsConvertible:
    return dst.(conversion.Convertible).ConvertFrom(src.(conversion.Hub))
case dstIsHub && srcIsConvertible:
    return src.(conversion.Convertible).ConvertTo(dst.(conversion.Hub))
case srcIsConvertible && dstIsConvertible:
    return p.convertViaHub(src, dst)
default:
    return fmt.Errorf("%T is not convertible to %T", dst, src)
}
```

DirectXMan12 commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,258 @@
+
+	var hub conversion.Hub
+	var isHub, hubFoundAlready bool
+	for _, gvk := range gvks {
```
we should probably cache this -- it's not going to change at runtime.

DirectXMan12 commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,258 @@
+
+// allocateDstObject returns an instance for a given GVK.
+func (wh *Webhook) allocateDstObject(apiVersion, kind string) (runtime.Object, error) {
+	gvk := schema.FromAPIVersionAndKind(apiVersion, kind)
```
don't we have a helper for this method lying around here somewhere?

DirectXMan12 replied:
could put this in a sync.Once, or something.  We can tackle in a follow-up

DirectXMan12 commented on pkg/webhook/conversion/conversion_suite_test.go:
```diff
@@ -0,0 +1,38 @@
+var _ = BeforeSuite(func(done Done) {
+	logf.SetLogger(zap.LoggerTo(GinkgoWriter, true))
+
+	close(done)
```
no need for a timeout if you're not launching a testenv

DirectXMan12 commented on pkg/webhook/conversion/conversion_test.go:
```diff
@@ -0,0 +1,227 @@
+
+	jobsapis "sigs.k8s.io/controller-runtime/pkg/webhook/conversion/testData/pkg/apis"
+	jobsv1 "sigs.k8s.io/controller-runtime/pkg/webhook/conversion/testData/pkg/apis/jobs/v1"
+	jobsv2 "sigs.k8s.io/controller-runtime/pkg/webhook/conversion/testData/pkg/apis/jobs/v2"
```
the future is now :puts on back to the future driving mask:

DirectXMan12 commented on pkg/webhook/conversion/conversion_test.go:
```diff
@@ -0,0 +1,227 @@
+
+		convReview := doRequest(convReq)
+
+		Expect(convReview.Response.ConvertedObjects).Should(HaveLen(1))
```
`Expect().To` (more idiomatic)

DirectXMan12 commented on pkg/webhook/conversion/conversion_test.go:
```diff
@@ -0,0 +1,227 @@
+		}
+
+		convReview := doRequest(convReq)
+		Expect(convReview.Response.Result.Message).Should(
```
blech.  I hate directly checking error message strings.  Can't we just check codes?

droot replied:
Need to explore a bit more, but in general I agree checking error messages is bad.

droot replied:
I explored bit more about using codes here, couldn't find an easy way to do it at this level which is conversionRequest/Response level. At this level, we can only indicate status and failure message (which contains the reason for failure). I removed reason string checking here because that makes the tests brittle.

DirectXMan12 commented on pkg/webhook/conversion/response.go:
```diff
@@ -0,0 +1,31 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func errored(err error) *apix.ConversionResponse {
```
just move this into the other file where it's used, probably

DirectXMan12 commented on pkg/webhook/conversion/testData/pkg/apis/addtoscheme_jobs_v1.go:
```diff
@@ -0,0 +1,25 @@
+/*
```
move this under `example/` (or just use `example/`), just to be certain we don't cause import issues

droot replied:
Done.

droot commented on pkg/conversion/conversion.go:
```diff
@@ -0,0 +1,34 @@
+limitations under the License.
+*/
+
+package conversion
```
Done.

droot commented on pkg/webhook/conversion/conversion.go:
```diff
@@ -0,0 +1,258 @@
+limitations under the License.
+*/
+
+package conversion
```
Done.

DirectXMan12 commented on pkg/webhook/conversion/conversion_test.go:
```diff
@@ -0,0 +1,203 @@
+		}
+
+		scheme = kscheme.Scheme
+		Expect(jobsapis.AddToScheme(scheme)).Should(Succeed())
```
nit: Expect().To
